align_system/algorithms/outlines_hybrid_regression_adm.py

```diff
@@ -0,0 +1,215 @@
+                                num_samples=1,
+                                predict_outcomes=False,
+                                distribution_matching='average',
+                                generator_batch_size=5,
+                                kdma_descriptions_map='align_system/prompt_engineering/kdma_descriptions.yml',
+                                kdma_score_examples=False,
+                                **kwargs):
+
+        scenario = []
+        for action in available_actions:
+            for character in scenario_state.characters:
+                chosen_character = None
+                if character.id == action.character_id:
+                    chosen_character = character
+            scenario_description = (
+                scenario_description_classic_regression(
+                    scenario_state,
+                    chosen_character
+                )
+            )
+            scenario.append(scenario_description)
+
+        # Important that the choices stay in the same order as the
+        # available actions as we'll use the selected index later to
+        # map to the corresponding action
+        choices = self.format_choices(
+            [a.unstructured for a in available_actions],
+            available_actions,
+            scenario_state
+        )
+
+        data_dict = {
+            'scenario': scenario,
+            'answer': choices
+        }
+
+        target_models = kwargs.get('models', {})
+        target_model_checkpoints = target_models["target_checkpoint"].keys()
+        model_name = target_models["model_name"]
+
+        target_kdmas = alignment_target.kdma_values
+
+        target_kdma_configs = []
+        for i, target_kdma in enumerate(target_kdmas):
+            target_kdma_name = target_kdma.kdma
+            if target_kdma_name in target_model_checkpoints:
+                log.info("Chosen KDMA: %s, and target value: %f", target_kdma_name, target_kdma.value)
+                predictions = get_kdma_predictions(
+                    model_name=model_name,
+                    data_dict=data_dict,
+                    model_path=target_models["target_checkpoint"][str(target_kdma_name)]
+                )
+                predictions = np.array([predictions])
+                dist_to_align_score = []
+                for pred in predictions:
+                    target_kdma_value = target_kdma.value
+                    dist = (target_kdma_value - pred)**2
+                    log.info("Distance to alignment target: %f", dist)
+                    dist_to_align_score.append(dist)
+
+                choice_index = np.argmin(dist_to_align_score)
+                final_choice = choices[choice_index]
+
+                target_kdma_configs.append(
+                    {
+                        "name": target_kdma_name,
+                        "score": predictions[choice_index] * 10,
+                        "action_choice": final_choice
+                    }
+                )
```

Comment:
So you've made changes on this in the latest version of the code, but I think there's some confusion still. One singular action and/or target (in the future) may have multiple KDMA values.

So we might be aligning to `{"ingroup_bias": 0.7, "moral_judgement": 0.2}` (at the same time), therefore we'd predict both two (in this case) KDMA values for each action and then select the action that minimizes the distance from the target overall, i.e. the singular action that is closest to **both** ingroup_bias=0.7 and moral_judgement=0.2. Does that make sense?

**However**, Jadie just pushed [alignment utilities](https://github.com/ITM-Kitware/align-system/blob/main/align_system/utils/alignment_utils.py) that I think you can make use of to avoid re-implementing this logic and also support distribution-based targets. She uses these in the comparative regression ADM [here](https://github.com/ITM-Kitware/align-system/blob/main/align_system/algorithms/outlines_regression_adm_comparative.py#L371-L397)

Reply:
@barry-ravichandran Let me know if you'd like help integrating [alignment utilities](https://github.com/ITM-Kitware/align-system/blob/main/align_system/utils/alignment_utils.py) with your ADM